Add item requests

Sweta304 opened the pull request.

Sweta304 committed:
feature:
- добавлен запрос вещи
- добавлена пагинация
- добавлены тесты

Sweta304 committed:
Merge branch 'main' of https://github.com/Sweta304/java-shareit into add-item-requests

alexache commented on src/main/java/ru/practicum/shareit/booking/dto/BookingDto.java:
```diff
@@ -0,0 +1,24 @@
+    private Long id;
+    private LocalDateTime start;
+    private LocalDateTime end;
+    private Item item;
```
ДТО объект, который хранит не-дто модели в себе, выглядит немного странно

alexache commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -101,36 +108,72 @@ public List<Booking> getAllBookings(Long bookerId, String rawState) throws UserN
         switch (rawState) {
             case "ALL":
-                bookings = bookingJpaRepository.findByBookerId(bookerId, sort);
+                if (from != null && size != null) {
```
Это выглядит общей часть для всех кейсов, ее стоит вынести до свитча, чтобы не дублировать столько раз

alexache commented on src/main/java/ru/practicum/shareit/requests/controller/ItemRequestController.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    @PostMapping
+    public ItemRequestDto addItemRequest(@RequestBody @Valid ItemRequest itemRequest, @RequestHeader("X-Sharer-User-Id") Long requestor) throws UserNotFoundException, RequestNotCorrectException {
```
Надо работать только с  дто в контроллере

alexache commented on src/main/java/ru/practicum/shareit/item/controller/ItemController.java:
```diff
@@ -44,13 +46,13 @@ public ItemWithBooking getItem(@PathVariable Long itemId, @RequestHeader("X-Shar
     @GetMapping("/search")
-    public List<ItemDto> searchItem(@RequestParam String text) {
-        return itemService.searchItem(text);
+    public List<ItemDto> searchItem(@RequestParam String text, @RequestParam(required = false) Integer from, @RequestParam(required = false) Integer size) throws PaginationNotCorrectException {
```
На аргументы from и size лучше написать значения по умолчанию. тогда и не придется поддерживать вариант, что они могут быть null, и код в сервисе станет почище

alexache commented on src/main/java/ru/practicum/shareit/requests/service/ItemRequestServiceImpl.java:
```diff
@@ -0,0 +1,96 @@
+    public ItemRequestDto addItemRequest(ItemRequest itemRequest, Long requestor) throws UserNotFoundException, RequestNotCorrectException {
+        User user = userJpaRepository.findById(requestor).orElseThrow(() -> new UserNotFoundException("Пользователя не существует"));
+
+        if (itemRequest.getDescription() == null
```
Такие проверки лучше иметь ближе к контроллеру

alexache commented on src/test/java/ru/practicum/shareit/booking/BookingControllerTest.java:
```diff
@@ -0,0 +1,172 @@
+@AutoConfigureMockMvc
+public class BookingControllerTest {
+    @MockBean
+    BookingService bookingService;
```
Не хватает модификаторов доступа у полей (в остальных тестовых классах тоже)

alexache commented on src/test/java/ru/practicum/shareit/item/ItemControllerTest.java:
```diff
@@ -0,0 +1,173 @@
+@AutoConfigureMockMvc
+public class ItemControllerTest {
+    @MockBean
+    BookingService bookingService;
```
Не очень понятно с этим тестом. Копипаст от теста на контроллер букинга?

Sweta304 replied:
да, скопипастила, а переписать забыла :(

Sweta304 committed:
refactor:
- для методов с пагинацией добавлены значения по умолчанию
- добавлены модификаторы доступа к полям в тестах
- исправлены замечания по коду
- добавлены тесты для класса ItemController
- добавлены тесты для класса BookingMapper

Sweta304 committed:
refactor:
- исправлены тесты
- удален класс BookingRepositoryTest - т.к. из BookingJpaRepository удалены нативные запросы

Sweta304 committed:
refactor:
- добавлены интеграционные тесты

Sweta304 committed:
удалены неиспользуемые импорты

alexache commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -3,7 +3,7 @@
+import ru.practicum.shareit.requests.model.ItemRequest;
 import ru.practicum.shareit.user.model.User;
 
 import javax.persistence.*;
```
Кстати по коду, который ниже  - @Data/@EqualsAndHashCode нельзя использовать в классах-моделях - тех, что хранятся в БД

alexache commented on src/test/java/ru/practicum/shareit/booking/ShareItIntegrationTest.java:
```diff
@@ -0,0 +1,144 @@
+
+@SpringBootTest
+@AutoConfigureMockMvc
+@AutoConfigureTestDatabase
```
Для полноценного интеграционного теста не хватает реальной постгрес БД, но пусть будет так
